resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/regression/Bug_297635.java

```diff
@@ -0,0 +1,147 @@
+ *
+ *  Contributors:
+ *     IBM Corporation - initial API and implementation
+ *     Alexander Kurtakov <akurtako@redhat.com> - Bug 459343
+ *******************************************************************************/
+package org.eclipse.core.tests.resources.regression;
+
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.util.function.Consumer;
+import org.eclipse.core.internal.resources.SaveManager;
+import org.eclipse.core.resources.ISaveContext;
+import org.eclipse.core.resources.ISaveParticipant;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.tests.harness.BundleTestingHelper;
+import org.eclipse.core.tests.resources.ResourceTest;
+import org.eclipse.core.tests.resources.content.ContentTypeTest;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleException;
+
+/**
+ * Tests regression of bug 297635
+ */
+public class Bug_297635 extends ResourceTest {
+	private Bundle saveParticipantBundle;
+
+	@Override
+	protected void setUp() throws Exception {
+		super.setUp();
+		saveParticipantBundle = SaveParticipant.installContainingBundleAndRegister();
+		saveFull();
+	}
+
+	@Override
+	protected void tearDown() throws Exception {
+		if (saveParticipantBundle != null) {
+			saveParticipantBundle.uninstall();
+		}
+		super.tearDown();
+	}
+
+	public void testCleanSaveStateBySaveParticipantOnSnapshotSave() throws Exception {
+		executeWithSaveManagerSpy(saveManagerSpy -> {
+			try {
+				saveSnapshot(saveManagerSpy);
+			} catch (CoreException e) {
+			}
+			verify(saveManagerSpy).forgetSavedTree(saveParticipantBundle.getSymbolicName());
+		});
+	}
+
+	private void saveFull() throws CoreException {
+		getWorkspace().save(true, getMonitor());
+	}
+
+	private void saveSnapshot(SaveManager saveManager) throws CoreException {
+		saveManager.save(ISaveContext.SNAPSHOT, true, null, getMonitor());
+	}
+
+	private void executeWithSaveManagerSpy(Consumer<SaveManager> executeOnSpySaveManager)
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+		IWorkspace workspace = getWorkspace();
+		String saveManagerFieldName = "saveManager";
+		SaveManager originalSaveManager = (SaveManager) getField(workspace, saveManagerFieldName);
+		SaveManager spySaveManager = spy(originalSaveManager);
+		try {
+			setField(workspace, saveManagerFieldName, spySaveManager);
+			executeOnSpySaveManager.accept(spySaveManager);
+		} finally {
+			setField(workspace, saveManagerFieldName, originalSaveManager);
+		}
+	}
+
+	private static Object getField(Object object, String fieldName)
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
```

Comment:
NIT: Unless you foresee some exception that can be handled in the future (in a `catch`-block), I would even go as far as throwing only `Exception` in this and in every other method in the class. This would discourage catching the exception and calling `fail` in the future.

Your call though.

Reply:
Fixed in 69978f2.